tortoisebeacon/handlers.go

```diff
@@ -22,95 +23,92 @@ const TBFirstVotingProtocol = "TBFirstVotingGossip"
 // TBFollowingVotingProtocol is a protocol for sending Tortoise Beacon following voting messages through Gossip.
 const TBFollowingVotingProtocol = "TBFollowingVotingGossip"
 
-var (
-	// ErrMalformedProposal is returned if proposal message is malformed.
-	ErrMalformedProposal = errors.New("malformed proposal message")
-)
+// ErrMalformedProposal is returned if proposal message is malformed.
+var ErrMalformedProposal = errors.New("malformed proposal message")
 
 // HandleSerializedProposalMessage defines method to handle Tortoise Beacon proposal Messages from gossip.
 func (tb *TortoiseBeacon) HandleSerializedProposalMessage(ctx context.Context, data service.GossipMessage, sync service.Fetcher) {
 	tb.Log.With().Debug("New proposal message",
 		log.String("from", data.Sender().String()))
 
-	var m ProposalMessage
-	if err := types.BytesToInterface(data.Bytes(), &m); err != nil {
+	var message ProposalMessage
+	if err := types.BytesToInterface(data.Bytes(), &message); err != nil {
 		tb.Log.With().Error("Received malformed proposal message",
 			log.String("message", string(data.Bytes())),
 			log.Err(err))
 
 		return
 	}
 
-	if err := tb.handleProposalMessage(ctx, m); err != nil {
-		tb.Log.With().Error("Failed to handle proposal message",
-			log.String("sender", data.Sender().String()),
-			log.String("message", m.String()),
-			log.Err(err))
+	currentEpoch := tb.currentEpoch()
+	if message.EpochID < currentEpoch {
+		tb.Log.With().Debug("Received proposal message from previous epoch, ignoring",
+			log.Uint64("message_epoch", uint64(message.EpochID)),
+			log.Uint64("current_epoch", uint64(currentEpoch)))
 
 		return
 	}
 
-	data.ReportValidation(ctx, TBProposalProtocol)
-}
+	tb.proposalChansMu.Lock()
+	ch := tb.ensureProposalChanExists(message.EpochID)
+	tb.proposalChansMu.Unlock()
 
-func (tb *TortoiseBeacon) handleProposalMessage(ctx context.Context, m ProposalMessage) error {
-	// TODO(nkryuchkov): buffer messages that came from future epochs because of time drifts on different nodes (https://github.com/spacemeshos/go-spacemesh/pull/2529#issuecomment-877772213)
+	extendedMessage := extendedProposalMessage{
+		message: message,
+		gossip:  data,
+	}
 
-	receivedTimestamp := time.Now()
+	go func() {
+		select {
+		case <-tb.CloseChannel():
+		case ch <- extendedMessage:
+		}
+	}()
```

Comment:
creating a goroutine just to push a message is a bad practice. it is a like an unlimited queue, which shouldn't be used. unlimited queue is probably even cheaper

if you can it is better to think what is a max number of pending messages can be allowed, and then decide what you will do with them (pop last, pop first, or return error for example)

Reply:
@dshulyak Well, it may look not that great, however, `HandleSerializedProposalMessage` is not run in a separate goroutine (https://github.com/spacemeshos/go-spacemesh/blob/feature/buffer-proposal-messages/p2p/service/gossip_listener.go#L83), so I don't want to block it. The channel is buffered and has some capacity, but I don't want either to drop messages that don't fit the buffer or to make the channel capacity extremely big.

Reply:
i think that you are creating an unlimited queue instead. i am pretty sure that this pattern must be avoided, but it is up to you

in my opinion, correct system behavior would be to block producer (peer that sends messages to us) until the queue has enough capacity to accept the message, do you know if we have a listener for each peer or a global one?

Reply:
proposal channel `ch` is a buffered channel of 1024 items, so there is some inherent buffering here
I think that since there can be many proposals gossiped it is better to block on writing to this channel until it is available, otherwise we might have a lot of goroutines pending
Lets consider however refactoring the gossip listener code to have a worker pool or some additional buffering